### PR TITLE
CPM-501: Validate valueIntents to avoid two edits on the same value

### PR DESCRIPTION
**Pull request by Balamung**
**Description (for Contributor and Core Developer)**

**Definition Of Done (for Core Developer only)**

- [ ] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Specification/Infrastructure/Validation/UserIntentsShouldBeUniqueValidatorSpec.php [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    function it_throws_an_exception_with_a_wrong_constraint()
+    {
+        $this->shouldThrow(\InvalidArgumentException::class)->during('validate', [1, new Type([])]);
+    }
+
+    function it_does_nothing_when_the_value_intents_are_distinct(ExecutionContext $context)
+    {
+        $context->addViolation(Argument::cetera())->shouldNotBeCalled();
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Product/back/Test/Specification/Infrastructure/Validation/UserIntentsShouldBeUniqueValidatorSpec.php:37]: addViolation is not directly called. We should test buildViolation instead

```suggestion
        $context->buildViolation(Argument::cetera())->shouldNotBeCalled();
```

**Review by nmarniesse — APPROVED** [review]

File: src/Akeneo/Pim/Enrichment/Product/back/Infrastructure/Validation/UserIntentsShouldBeUniqueValidator.php [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        /** @var ValueUserIntent $valueUserIntent */
+        foreach ($valueUserIntents as $valueUserIntent) {
+            $intentLocale = $valueUserIntent->localeCode() ?? '<all_locales>';
+            $intentChannel = $valueUserIntent->channelCode() ?? '<all_channels>';
+            $intentAttributeCode = $valueUserIntent->attributeCode();
+
+            if (\array_key_exists($intentLocale, $existingIntents) &&
+                \array_key_exists($intentChannel, $existingIntents[$intentLocale]) &&
+                \in_array($intentAttributeCode, $existingIntents[$intentLocale][$intentChannel])
+            ) {
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Product/back/Infrastructure/Validation/UserIntentsShouldBeUniqueValidator.php:35]:
```suggestion
    if (\in_array($intentAttributeCode, $existingIntents[$intentLocale][$intentChannel] ?? [])) {
```

**Review by mmetayer — APPROVED** [review]
Shouldn't we also validate the unicity of other user intents (i.e: only one intent on categories, only one on status, only one on family, etc.) ?

File: src/Akeneo/Pim/Enrichment/Product/back/Infrastructure/Validation/UserIntentsShouldBeUniqueValidator.php (hunk 2 above)
Comment (Balamung) [reply src/Akeneo/Pim/Enrichment/Product/back/Infrastructure/Validation/UserIntentsShouldBeUniqueValidator.php]: I wasn't sure how to handle those as they don't exist yet, any idea ? :)

**Comment by nmarniesse** [comment]
> 

@mmetayer we cannot have multiple intents for family/category/... Just for value user intents

```
// UpsertProductCommand.php

public function __construct(
        private int $userId,
        private string $productIdentifier,
        private mixed $identifierUserIntent = null,
        private mixed $familyUserIntent = null,
        private mixed $categoryUserIntent = null,
        private mixed $parentUserIntent = null,
        private mixed $groupsUserIntent = null,
        private mixed $enabledUserIntent = null,
        private mixed $associationsUserIntent = null,
        private array $valueUserIntents = []
    ) {
```

**Comment by mmetayer** [comment]
You're right, I'm feeling stupid tight now :sweat_smile:

**Pull request merged**